### ISSUE TITLE
Implement `http2`

### ISSUE THREAD

**Issue by jgunzelman88** [oven-sh/bun#887]
### Version

0.1.5

### Platform

Darwin ip-192-168-1-189.us-east-2.compute.internal 21.5.0 Darwin Kernel Version 21.5.0: Tue Apr 26 21:08:22 PDT 2022; root:xnu-8020.121.3~4/RELEASE_X86_64 x86_64

### What steps will reproduce the bug?

Don't know if this is just a module that hasn't been added yet but I am trying to use the GCP datastore apk and I got

error: Could not resolve: "http2". Maybe you need to "bun install"?

I did bun add http2 and this was resolved but then I get

error: Cannot find module "./framer" from "bun:wrap"

### How often does it reproduce? Is there a required condition?

https://github.com/poketrax/pokepull/tree/v1.0.0

bun ./src/pokepull.ts

Or install the gcp datastore apk
bun install @google-cloud/datastore
instantiate a Datastore object.

### What is the expected behavior?

Module should resolve.

### What do you see instead?

error: Cannot find module "./framer" from "bun:wrap"

### Additional information

Let me know if this is something I need to submit to GCP or if I did something stupid

**Comment by thiagozanetti** [comment]
Same behavior happening here but my issue is with GRPC from etcd3.

**Comment by jgunzelman88** [comment]
So bug is gone now in 0.1.7 but now i get a segmentation fault :/

**Comment by jgunzelman88** [comment]
its a seg fault at 0 so probably a null pointer error but i'm no zig expert.
`SegmentationFault at 0x0000000000000000`

**Comment by imadbz** [comment]
same issue with "firebase-admin".
http2 not found, when I "bun add" it, I get 
"error: Cannot find module "./framer" from "bun:wrap"

bun v0.1.9

**Comment by jgunzelman88** [comment]
I am now getting "error: Cannot find module "./framer" from "bun:wrap" again

bun v0.1.10

**Comment by irdevp** [comment]
error: Cannot find module "./framer" from "bun:wrap"
v0.1.10

**Comment by Electroid** [comment]
We do not support the Node.js `http2` module yet, but are tracking support for it.

**Comment by evelant** [comment]
I just discovered bun and got hit with this on the very first script I tried to run 😦 

The script talks to firebase services so uses grpc so requires http2. Back to ts-node for now 🤷

**Comment by AndreaCuneo** [comment]
Same here: gRPC and gRPC-web is getting traction in new projects, firebase in particular, and bun is grand for new projects. Also this hits with a segmentation fault which takes time to diagnose to this issue..

**Comment by lewisl9029** [comment]
Looks like `@aws-sdk/client-s3` is also blocked by this. For S3 I can probably just replace with something fetch-based, but could be more problematic for other aws services.

**Comment by tsp1998** [comment]
I am getting http2 error for googleapis

Error: Cannot find package "http2" from "/home/royaltsp/tsp/node/google-drive-files/google-drive-demo/node_modules/googleapis-common/build/src/http2.js"

Code: https://github.com/tsp1998/google-drive-demo

**Comment by NyanHelsing** [comment]
kind of a big blocker for adoption. I literally can't solve my problem using bun so now I need to rewrite for deno or node.js which is a massive drag.

**Comment by ph1losof** [comment]
Massive block for me

**Comment by VrindavanSanap** [comment]
Yeah me too firebase does not work due to this

**Comment by birkskyum** [comment]
Still issue in 0.7.1 - @Jarred-Sumner will this be in the 1.0 release?

**Comment by prosenjitjoy** [comment]
v0.8.1 - getting this error while starting a gRPC server

**Comment by c0dification** [comment]
v1.0.0 still an issue.

**Comment by Electroid** [comment]
This is at the top of our feature list, @cirospaciari has already started work on this.

**Comment by nostrascott** [comment]
Would love to see this, as I love bun, but also need to use the AWS SDK.  Cheers to all those working hard on bun!

**Comment by o-alexandrov** [comment]
`bun` is amazing and extremely valuable to the community, thank you.
However, reading the following from [Bun v1 announcement](https://bun.sh/blog/bun-v1.0) was hard w/o a smile given `@aws-sdk/*` and many other packages depending on this issue don't work.

> While perfect compatibility with Node.js isn't possible — looking at you node:v8 — Bun can run virtually any Node.js application in the wild.

**Comment by x1957** [comment]
I really hope to be able to use bun, but I'm stuck with gRPC. Is there a place where I can see the progress of this feature's support?

**Comment by birkskyum** [comment]
@x1957 , this is that place. As mentioned above, work is already started on this.

**Comment by vsly-ar** [comment]
v1.0 released with this issue? Big bunmer.

**Comment by tsp1998** [comment]
Hope to see this in bun sooner!

**Comment by RahulDey12** [comment]
Please address this issue

**Comment by gartz** [comment]
Proto buffers are super common nowadays and every lib is taking advantage of http2 on their implementations.
Not having support to http2 will block any project that is integrating with Google’s infrastructure as they all rely on proto buffers.
Please prioritize adding http2 support to unblock migration from GCP consumers.
I’m looking forward to use Bun, as soon as I can migrate my micro-servers that all internally also communicate using proto buffers / gRPC.

**Comment by naterkane** [comment]
👍 +1

I am using a wrapped/fixed copy of the long-abandoned `spdy`, similar to many others. 
Though `bun` has been fantastic for package management, building, and test-running for CI. I can't use it as a runtime for the application itself due to this issue.

**Comment by moderation** [comment]
Deno's native HTTP/2 support documented at https://deno.land/manual@v1.36.4/runtime/http_server_apis#http2-support-1

**Comment by vrn-dev** [comment]
Will we see HTTP/2 released in v1.0.4?

**Comment by birkskyum** [comment]
@vrn-dev , there was a [mention](https://youtu.be/pOuR61Ywg14?feature=shared&t=1929) two weeks ago, that it was expected in a week or two, so it might be close.

**Comment by dbousamra** [comment]
@birkskyum the branch hasn't made any progress in weeks so seems unlikely?: https://github.com/oven-sh/bun/tree/ciro/http2

**Comment by avpengage** [comment]
any updates?  my bun project is at a standstill...

Looking at the branch mentioned above doesn't show any commits for over a month...

**Comment by birkskyum** [comment]
@dbousamra , yeah, it's hopefully next up soon, but it does indeed look stale.

**Comment by Jarred-Sumner** [comment]
@cirospaciari is working on it

**Comment by monkfromearth** [comment]
Any updates on this? This is a critical requirement, before going to production.

**Comment by Tanguille** [comment]
Any alternatives/solutions? This is preventing me to use bun for my existing react PWA

**Comment by Zacaria** [comment]
@cirospaciari is working on it => https://github.com/oven-sh/bun/compare/ciro/http2

**Comment by hugo082** [comment]
Based on the changes I see it plan to fix the [got issue](https://github.com/sindresorhus/got/issues/2299) too right?

**Comment by netaisllc** [comment]
This one also seems to be blocking use of **Bun** with **FaunaDB JS SDK**.  I have a POC script that gives this failure:

```
                            ^
NotImplementedError: node:http2 connect is not yet implemented in Bun. Track the status & thumbs up the issue: https://github.com/oven-sh/bun/issues/887
 code: "ERR_NOT_IMPLEMENTED"

      at new NotImplementedError (internal:shared:13:26)
      at internal:shared:2:68
      at node:http2:2:44
      at connect_fn (/Users/kevinm/Projects/Strongcloud/fauna-lab/issue_6204/node_modules/fauna/dist/node/index.js:420:24)
      at /Users/kevinm/Projects/Strongcloud/fauna-lab/issue_6204/node_modules/fauna/dist/node/index.js:461:22
      at new Promise (:1:20)
      at doRequest_fn (/Users/kevinm/Projects/Strongcloud/fauna-lab/issue_6204/node_modules/fauna/dist/node/index.js:437:9)
      at /Users/kevinm/Projects/Strongcloud/fauna-lab/issue_6204/node_modules/fauna/dist/node/index.js:368:21
      at request (/Users/kevinm/Projects/Strongcloud/fauna-lab/issue_6204/node_modules/fauna/dist/node/index.js:363:16)
      at /Users/kevinm/Projects/Strongcloud/fauna-lab/issue_6204/node_modules/fauna/dist/node/index.js:1459:27

  } ]
```

**Comment by treeder** [comment]
Also seems to block usage of all the firebase admin packages: https://github.com/firebase/firebase-admin-node/issues/2300

**Comment by rupurt** [comment]
Looks like the initial http2 branch just got merged :tada: https://github.com/oven-sh/bun/pull/6761

I guess it'll drop in `1.0.13`

**Comment by naterkane** [comment]
well, partially drop. it doesn't look ready for action if you read the PR's description. i know it won't include enough support for my use case (unfortunately)

**Comment by birkskyum** [comment]
Was excited to try the new 1.0.13 release that claims _"http2 module work now"_.

<img width="449" alt="Screenshot 2023-11-18 at 16 32 55" src="https://github.com/oven-sh/bun/assets/74932975/3dfc0a54-9bac-47f9-a673-170050525372">


I still see this though:

```sh
error when starting dev server:
NotImplementedError: node:http2 createSecureServer is not yet implemented in Bun. Track the status & thumbs up the issue: https://github.com/oven-sh/bun/issues/887
    at new NotImplementedError (internal:shared:13:14)
    at <anonymous> (internal:shared:2:35)
    at <anonymous> (node:http2:50:33)
    at processTicksAndRejections (:61:39)
```

That said - Great to see it progress! 👏

**Comment by debkanchan** [comment]
Great to see http2 finally land. Was excited to try it but faced the following error
```bash
NotImplementedError: node:http2 createServer is not yet implemented in Bun. Track the status & thumbs up the issue: https://github.com/oven-sh/bun/issues/887
```

**Comment by rupurt** [comment]
It's client only for now @DebkanchanSamadder. Very much looking forward to the server!

**Comment by marc-barry** [comment]
Am I correct in understanding that with https://bun.sh/blog/bun-v1.0.13#http2-client-support that only the client side is implemented at this time. I see that TLS on the server side is implemented and available under https://bun.sh/docs/api/http#tls but is it that there is no HTTP/2 support at all (natively) in Bun?

**Comment by birkskyum** [comment]
@marc-barry , yes, the http2.createServer and http2.createSecureServer aren't implemented yet, so only client for now

**Comment by deathemperor** [comment]
This prevents running vite with Bun in development when server.https is used. Looking forward for this support.

**Comment by elzorrorebelde** [comment]
beware: you won't be able to write a gRPC server with bun until this get resolved

**Comment by MarcStdt** [comment]
Is there any ETA for the server part?
I could not find anything in the roadmap.

**Comment by dgrbrady** [comment]
> beware: you won't be able to write a gRPC server with bun until this get resolved

Man I wish this was more clear in the [release notes](https://bun.sh/blog/bun-v1.0.13#grpc-works-now). I was super excited to implement my gRPC server in bun, but when I run my `.ts` file, it just quits the process and doesn't stay open.

**Comment by vsly-ar** [comment]
@dgrbrady @MarcStdt @deathemperor  I've opened a new issue #8823, focused solely on the server side of HTTP2 support. Judging by the release notes, the core team chooses features to implement based on the number of upvotes for an issue. The upvotes on this issue may already be considered "solved." Therefore, I believe we need to upvote the new one to make progress on the matter.

**Comment by Electroid** [comment]
Thanks @vsly-ru for opening that issue. Our intention was to keep this issue open until we support the HTTP2 server as well, but it's probably a good idea that we use a new issue to get more realistic upvotes.

To be clear, we do plan to support the HTTP2 server. If that's something you're depending on, please upvote the issue: https://github.com/oven-sh/bun/issues/8823

If you only need the HTTP2 client, as of Bun v1.0.13 it is supported.

**Comment by birkskyum** [comment]
Thanks @vsly-ru for making a dedicated ticket for http2 server part:
- https://github.com/oven-sh/bun/issues/8823

**Comment by guest271314** [comment]
> If you only need the HTTP2 client, as of Bun v1.0.13 it is supported.

@Electroid  The HTTP client doesn't work.

**Comment by ahmadnasriya** [comment]
I'm using Bun v1.1.4, and the issue still persisting.

```shell
NotImplementedError: node:http2 createSecureServer is not yet implemented in Bun. Track the status & thumbs up the issue: https://github.com/oven-sh/bun/issues/887
```

How is this issue closed? I see some comments in this artivle claiming that it's working on earlier versions, but how come!? It's been five months now

**Comment by birkskyum** [comment]
@nasriyasoftware this ticket was closed when http2 client was implemented, the title just wasn't updated. The server issue is still open, and the most upvoted open ticket in bun repo at this point.

- https://github.com/oven-sh/bun/issues/8823

**Comment by ahmadnasriya** [comment]
@birkskyum thank you ;)
